File: SRP/userregistration.java
class User {
    private String name;
    private String email;

    public User(String name, String email) {
        this.name = name;
        this.email = email;
    }

    public String getName() { 
        return name; 
    }
    public String getEmail() { 
        return email; 
    }
}

class UserValidator {
    public boolean isValid(User user) {
        return user.getEmail().contains("@") && !user.getName().isEmpty();
    }
}

class UserRepository {
    public void save(User user) {
        System.out.println("Saving user " + user.getEmail() + " to database");
    }
}

class EmailService {
    public void sendWelcomeMail(User user) {
        System.out.println("Sending welcome email to " + user.getEmail());
    }
}


public class userregistration {
    public static void main(String[] args) {

        User user = new User("abc", "abc@mail.com");
        UserValidator vali = new UserValidator();
        UserRepository repo = new UserRepository();
        EmailService email = new EmailService();

        if (vali.isValid(user)) {
            repo.save(user);
            email.sendWelcomeMail(user);
        } 
        else {
            System.out.println("Invalid user data");
        }
    }
}

File: SRP/userregistration.ts
class User {
  constructor(private name: string, private email: string) {
  }

  getName(): string {
    return this.name;
  }

  getEmail(): string {
    return this.email;
  }
}

class UserValidator {
  isValid(user: User): boolean {
    return user.getEmail().includes("@") && user.getName().length > 0;
  }
}

class UserRepository {
  save(user: User): void {
    console.log(`Saving user ${user.getEmail()} to database`);
  }
}

class EmailService {
  sendWelcomeEmail(user: User): void {
    console.log(`Sending welcome email to ${user.getEmail()}`);
  }
}

const user = new User("abc", "abc@mail.com");

const vali = new UserValidator();
const repo = new UserRepository();
const email = new EmailService();

if (vali.isValid(user)) {
  repo.save(user);
  email.sendWelcomeEmail(user);
} else {
  console.log("Invalid user data");
}